src/heartbeats.cpp

```diff
@@ -0,0 +1,611 @@
+/**
+ * @file heartbeats.cpp
+ * @brief Classes for heartbeating Sync configuration and status
+ *
+ * (c) 2013 by Mega Limited, Auckland, New Zealand
+ *
+ * This file is part of the MEGA SDK - Client Access Engine.
+ *
+ * Applications using the MEGA API must present a valid application key
+ * and comply with the the rules set forth in the Terms of Service.
+ *
+ * The MEGA SDK is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *
+ * @copyright Simplified (2-clause) BSD License.
+ *
+ * You should have received a copy of the license along with this
+ * program.
+ */
+
+#include "mega/heartbeats.h"
+#include "mega/command.h"
+#include "assert.h"
+
+namespace mega {
+
+
+HeartBeatSyncInfo::HeartBeatSyncInfo(int tag, handle id)
+    : mHeartBeatId(id), mSyncTag(tag)
+{
+}
+
+handle HeartBeatSyncInfo::heartBeatId() const
+{
+    return mHeartBeatId;
+}
+
+Command *HeartBeatSyncInfo::runningCommand() const
+{
+    return mRunningCommand;
+}
+
+void HeartBeatSyncInfo::setRunningCommand(Command *runningCommand)
+{
+    mRunningCommand = runningCommand;
+}
+
+HeartBeatSyncInfo::Status HeartBeatSyncInfo::status() const
+{
+    return mStatus;
+}
+
+double HeartBeatSyncInfo::progress() const
+{
+    return std::max(0., std::min(1., static_cast<double>(mTransferredBytes) / static_cast<double>(mTotalBytes)));
+}
+
+uint32_t HeartBeatSyncInfo::pendingUps() const
+{
+    return mPendingUps;
+}
+
+uint32_t HeartBeatSyncInfo::pendingDowns() const
+{
+    return mPendingDowns;
+}
+
+m_time_t HeartBeatSyncInfo::lastAction() const
+{
+    return mLastAction;
+}
+
+mega::MegaHandle HeartBeatSyncInfo::lastSyncedItem() const
+{
+    return mLastSyncedItem;
+}
+
+void HeartBeatSyncInfo::updateTransferInfo(MegaTransfer *transfer)
+{
+    if (mPendingTransfers.find(transfer->getTag()) == mPendingTransfers.end())
+    {
+        mPendingTransfers.insert(std::make_pair(transfer->getTag(), ::mega::make_unique<PendingTransferInfo>()));
+    }
+
+    const auto &pending = mPendingTransfers[transfer->getTag()];
```

Comment:
We should avoid the double lookup in the map. It should be fast, but we know that it might become a bottleneck easily when there are a huge number of transfers already in the queue. Better to have an `if-else`. Something like:
```suggestion cpp
    const auto it = mPendingTransfers.find(transfer->getTag());
    PendingTransferInfo *pending = nullptr;
    if (it == mPendingTransfers.end())
    {
        pending = mPendingTransfers.insert(std::make_pair(transfer->getTag(), ::mega::make_unique<PendingTransferInfo>())).first->second.get();
    }
    else
        pending = it->second.get();
```

Reply:
I have it on a stash for this method and the `removePendingTransfer()`, let me know if you agree.

Reply:
Solved